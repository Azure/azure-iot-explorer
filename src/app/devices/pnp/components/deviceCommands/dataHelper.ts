/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { ModelDefinition } from '../../../../api/models/modelDefinition';
import { JsonSchemaAdaptor } from '../../../../shared/utils/jsonSchemaAdaptor';
import { CommandSchema } from './deviceCommandsPerInterfacePerCommand';

interface DeviceCommandsWithSchema {
    commandSchemas: CommandSchema[];
}

export const getDeviceCommandPairs = (modelDefinition: ModelDefinition): DeviceCommandsWithSchema => {
    const jsonSchemaAdaptor = new JsonSchemaAdaptor(modelDefinition);
    const commands = jsonSchemaAdaptor.getCommands();
    return {
        // tslint:disable-next-line: no-any
        commandSchemas: commands.map((command: any) => ({
            commandModelDefinition: command,
            parsedSchema: jsonSchemaAdaptor.parseInterfaceCommandToJsonSchema(command),
        }))
    };
};
