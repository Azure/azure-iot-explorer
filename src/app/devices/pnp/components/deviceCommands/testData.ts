/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { ModelDefinition } from '../../../../api/models/modelDefinition';
import { PnpStateInterface, pnpStateInitial } from '../../state';
import { SynchronizationStatus } from '../../../../api/models/synchronizationStatus';
import { REPOSITORY_LOCATION_TYPE } from '../../../../constants/repositoryLocationTypes';

const interfaceId = 'urn:contoso:com:environmentalsensor;1';
/* tslint:disable */
export const testModelDefinition: ModelDefinition = {
    "@id": interfaceId,
    "@type": "Interface",
    "displayName": "Digital Twin",
    "contents": [
        {
            "@type": "Command",
            "description": "This command will begin blinking the LED for given time interval.",
            "name": "blink",
            "request": {
                "name": "blinkRequest",
                "schema": "long"
            },
            "response": {
                "name": "blinkResponse",
                "schema": "string"
            },
            "commandType": "synchronous"
        }
    ],
    "@context": "https://azureiot.com/v1/contexts/Interface.json"
};

export const pnpStateWithTestData: PnpStateInterface = {
    ...pnpStateInitial(),
    modelDefinitionWithSource: {
        synchronizationStatus: SynchronizationStatus.fetched,
        payload: {
            isModelValid: true,
            modelDefinition: testModelDefinition,
            source: REPOSITORY_LOCATION_TYPE.Public
        }
    }
}