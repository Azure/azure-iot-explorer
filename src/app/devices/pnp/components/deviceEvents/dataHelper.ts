/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { ModelDefinition, TelemetryContent } from '../../../../api/models/modelDefinition';
import { JsonSchemaAdaptor } from '../../../../shared/utils/jsonSchemaAdaptor';
import { ParsedJsonSchema } from '../../../../api/models/interfaceJsonParserOutput';

export interface TelemetrySchema {
    parsedSchema: ParsedJsonSchema;
    telemetryModelDefinition: TelemetryContent;
}

export const getDeviceTelemetry = (modelDefinition: ModelDefinition): TelemetrySchema[] => {
    if (!modelDefinition) {
        return [];
    }
    const jsonSchemaAdaptor = new JsonSchemaAdaptor(modelDefinition);
    const telemetryContents = jsonSchemaAdaptor.getTelemetry();
    return telemetryContents.map(telemetry => ({
        parsedSchema: jsonSchemaAdaptor.parseInterfaceTelemetryToJsonSchema(telemetry),
        telemetryModelDefinition: telemetry
    }));
};
