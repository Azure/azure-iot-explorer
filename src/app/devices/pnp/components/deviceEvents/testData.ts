/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { ModelDefinition } from '../../../../api/models/modelDefinition';

const interfaceId = 'urn:azureiot:samplemodel;1';
export const testComponentName = 'environmentalSensor';
/* tslint:disable */
export const testModelDefinition: ModelDefinition = {
    "@id": interfaceId,
    "@type": "Interface",
    "displayName": "Digital Twin",
    "contents": [
        {
            "@type": "Telemetry",
            "name": "humid",
            "displayName": "Temperature",
            "description": "Current temperature on the device",
            "schema": "double"
        },
    ],
    "@context": "https://azureiot.com/v1/contexts/Interface.json"
}
/* tslint:enable */
