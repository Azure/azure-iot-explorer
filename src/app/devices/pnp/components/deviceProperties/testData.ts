/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { ModelDefinition } from '../../../../api/models/modelDefinition';
import { PnpStateInterface, pnpStateInitial } from '../../state';
import { SynchronizationStatus } from '../../../../api/models/synchronizationStatus';
import { REPOSITORY_LOCATION_TYPE } from '../../../../constants/repositoryLocationTypes';

const interfaceId = 'urn:azureiot:samplemodel;1';
export const testComponentName = 'environmentalSensor';
/* tslint:disable */
export const modelInformationReportedValue = {
    "modelId": "urn:contoso:com:dcm:2",
    "interfaces": {
        "urn_azureiot_ModelDiscovery_ModelInformation": "urn:azureiot:ModelDiscovery:ModelInformation:1",
        componentName: interfaceId
    }
};

export const testTwin: any = {
    "deviceId": "testDevice",
    "$modelId": interfaceId,
    properties: {
        reported: {
            modelInformation: modelInformationReportedValue
        }
    }
};

testTwin.properties.reported[testComponentName] = {
    "modelInformation": modelInformationReportedValue,  // component level model information
    "__t": "c"
};

const sampleSenmanticProperty = {
    '@type': [
        "Property",
        "SemanticType/Brightness Level"
        ],
    displayName: 'Brightness Level',
    description: 'The brightness level for the light on the device. Can be specified as 1 (high), 2 (medium), 3 (low)',
    name: 'brightness',
    schema: 'long'
};

export const testModelDefinition: ModelDefinition = {
    "@id": interfaceId,
    "@type": "Interface",
    "displayName": "Digital Twin",
    "contents": [
        {
            "@type": "Property",
            "name": "modelInformation",
            "displayName": "Model Information",
            "description": "Providing model and optional interfaces information on a digital twin.",
            "schema": {
                "@type": "Object",
                "fields": [
                    {
                        "name": "modelId",
                        "schema": "string"
                    },
                    {
                        "name": "interfaces",
                        "schema": {
                            "@type": "Map",
                            "mapKey": {
                                "name": "name",
                                "schema": "string"
                            },
                            "mapValue": {
                                "name": "schema",
                                "schema": "string"
                            }
                        }
                    }
                ]
            }
        },
        sampleSenmanticProperty
    ],
    "@context": "https://azureiot.com/v1/contexts/Interface.json"
}

export const pnpStateWithTestData: PnpStateInterface = {
    ...pnpStateInitial(),
    modelDefinitionWithSource: {
        synchronizationStatus: SynchronizationStatus.fetched,
        payload: {
            modelDefinition: testModelDefinition,
            source: REPOSITORY_LOCATION_TYPE.Local,
            isModelValid: true
        },
    },
    twin: {
        payload: testTwin,
        synchronizationStatus: SynchronizationStatus.fetched
    }
}
