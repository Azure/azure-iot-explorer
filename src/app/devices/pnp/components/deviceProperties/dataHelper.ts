/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { ModelDefinition, PropertyContent } from '../../../../api/models/modelDefinition';
import { JsonSchemaAdaptor } from '../../../../shared/utils/jsonSchemaAdaptor';
import { ParsedJsonSchema } from '../../../../api/models/interfaceJsonParserOutput';
import { ModuleTwin } from '../../../../api/models/moduleTwin';
import { Twin } from '../../../../api/models/device';
import { getReportedPropertiesForSpecficComponent } from '../deviceSettings/dataHelper';

export interface TwinWithSchema {
    propertyModelDefinition: PropertyContent;
    propertySchema: ParsedJsonSchema;
    reportedTwin: boolean | string | number | object;
}

export const generateReportedTwinSchemaAndInterfaceTuple = (model: ModelDefinition, twin: Twin | ModuleTwin, componentName: string): TwinWithSchema[] => {
    if (!model) {
        return [];
    }
    const jsonSchemaAdaptor = new JsonSchemaAdaptor(model);
    const nonWritableProperties = jsonSchemaAdaptor.getNonWritableProperties();

    return nonWritableProperties.map(property => ({
        propertyModelDefinition: property,
        propertySchema: jsonSchemaAdaptor.parseInterfacePropertyToJsonSchema(property),
        reportedTwin: getReportedPropertiesForSpecficComponent(twin, componentName)?.[property.name]
    }));
};
