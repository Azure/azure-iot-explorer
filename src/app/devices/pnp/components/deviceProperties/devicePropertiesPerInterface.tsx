/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { DetailsListLayoutMode, IColumn, CheckboxVisibility, Label, Overlay, ActionButton, Panel, PanelType } from '@fluentui/react';
import { ResizableDetailsList } from '../../../../shared/resizeDetailsList/resizableDetailsList';
import { ResourceKeys } from '../../../../../localization/resourceKeys';
import { getLocalizedData } from '../../../../api/dataTransforms/modelDefinitionTransform';
import { RenderSimplyTypeValue } from '../../../shared/components/simpleReportedSection';
import { ComplexReportedFormPanel } from '../../../shared/components/complexReportedFormPanel';
import { SemanticUnit } from '../../../../shared/units/components/semanticUnit';
import { TwinWithSchema } from './dataHelper';
import { getSchemaType, isSchemaSimpleType } from '../../../../shared/utils/jsonSchemaAdaptor';

export interface DevicePropertiesDataProps {
    twinAndSchema: TwinWithSchema[];
}

export const DevicePropertiesPerInterface: React.FC<DevicePropertiesDataProps> = (props: DevicePropertiesDataProps) => {
    const { t } = useTranslation();
    const [showOverlay, setShowOverlay] = React.useState<boolean>(false);
    const [showReportedValuePanel, setShowReportedValuePanel] = React.useState<boolean>(false);
    const [selectedItem, setSelectedItem] = React.useState(undefined);

    const getColumns = (): IColumn[] => {
        return [
            { key: 'name', name: t(ResourceKeys.deviceProperties.columns.name), fieldName: 'name', minWidth: 100, isMultiline: true },
            { key: 'schema', name: t(ResourceKeys.deviceProperties.columns.schema), fieldName: 'schema', minWidth: 100, isMultiline: true },
            { key: 'unit', name: t(ResourceKeys.deviceProperties.columns.unit), fieldName: 'unit', minWidth: 100, isMultiline: true },
            { key: 'value', name: t(ResourceKeys.deviceProperties.columns.value), fieldName: 'value', minWidth: 150, isMultiline: true }
        ];
    };

    const renderItemColumn = (item: TwinWithSchema, index: number, column: IColumn) => {
        switch (column.key) {
            case 'name':
                return renderPropertyName(item);
            case 'schema':
                return renderPropertySchema(item);
            case 'unit':
                return renderPropertyUnit(item);
            case 'value':
                return renderPropertyReportedValue(item);
            default:
                return <></>;
        }
    };

    const renderPropertyName = (item: TwinWithSchema) => {
        const ariaLabel = t(ResourceKeys.deviceProperties.columns.name);
        let displayName = getLocalizedData(item.propertyModelDefinition.displayName);
        displayName = displayName ? displayName : '--';
        let description = getLocalizedData(item.propertyModelDefinition.description);
        description = description ? description : '--';
        return <Label aria-label={ariaLabel}>{item.propertyModelDefinition.name} ({displayName} / {description})</Label>;
    };

    const renderPropertySchema = (item: TwinWithSchema) => {
        const ariaLabel = t(ResourceKeys.deviceProperties.columns.schema);
        const propertyModelDefinition = item.propertyModelDefinition;
        const schemaType = getSchemaType(propertyModelDefinition.schema);
        return <Label aria-label={ariaLabel}>{schemaType}</Label>;
    };

    const renderPropertyUnit = (item: TwinWithSchema) => {
        const ariaLabel = t(ResourceKeys.deviceProperties.columns.unit);
        return (
            <Label aria-label={ariaLabel}>
                <SemanticUnit unitHost={item.propertyModelDefinition} />
            </Label>);
    };

    // tslint:disable-next-line:cyclomatic-complexity
    const renderPropertyReportedValue = (item: TwinWithSchema) => {
        if (!item) {
            return <></>;
        }
        const ariaLabel = t(ResourceKeys.deviceProperties.columns.value);
        // tslint:disable-next-line:no-any
        const reportedValue = (item.reportedTwin as any)?.value || item.reportedTwin;
        // tslint:disable-next-line:no-any
        let displayValue = (item.propertyModelDefinition.schema as any)?.enumValues?.find((value: { enumValue: any; }) => value.enumValue === reportedValue)?.displayName || reportedValue;
        // tslint:disable-next-line:no-any
        displayValue = (displayValue as any)?.en || displayValue;
        return (
            <div aria-label={ariaLabel}>
                {
                    item.propertySchema && isSchemaSimpleType(item.propertyModelDefinition.schema, item.propertySchema.$ref) ?
                        RenderSimplyTypeValue(
                            item.reportedTwin,
                            item.propertySchema,
                            displayValue,
                            t(ResourceKeys.deviceProperties.columns.error)) :
                        item.reportedTwin ?
                            <ActionButton
                                className="column-value-button"
                                ariaDescription={t(ResourceKeys.deviceProperties.command.openReportedValuePanel)}
                                onClick={onViewReportedValue(item)}
                            >
                                {t(ResourceKeys.deviceProperties.command.openReportedValuePanel)}
                            </ActionButton> : <Label>--</Label>
                }
            </div>
        );
    };

    const onViewReportedValue = (item: TwinWithSchema) => () => {
        setSelectedItem(item);
        setShowOverlay(true);
        setShowReportedValuePanel(true);
    };

    const createReportedValuePanel = () => {
        if (!selectedItem) {
            return <></>;
        }
        const { reportedTwin, propertyModelDefinition: modelDefinition, propertySchema: schema } = selectedItem;
        return (
            <Panel type={PanelType.medium} isOpen={showReportedValuePanel} isLightDismiss={true}>
                    <ComplexReportedFormPanel
                        showPanel={showReportedValuePanel}
                        formData={reportedTwin}
                        handleDismiss={removeOverlay}
                        schema={schema}
                        modelDefinition={modelDefinition}
                    />
            </Panel>
        );
    };

    const removeOverlay = () => {
        setSelectedItem(undefined);
        setShowOverlay(false);
    };

    return (
        <div className="pnp-detail-list scrollable-lg ms-Grid">
            <div className="list-detail pnp-properties">
                <ResizableDetailsList
                    checkboxVisibility={CheckboxVisibility.hidden}
                    onRenderItemColumn={renderItemColumn}
                    items={props.twinAndSchema}
                    columns={getColumns()}
                    layoutMode={DetailsListLayoutMode.justified}
                />
                {showOverlay && <Overlay />}
                {createReportedValuePanel()}
            </div>
        </div>
    );
};
