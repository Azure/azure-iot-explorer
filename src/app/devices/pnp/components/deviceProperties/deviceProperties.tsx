/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { useLocation, useHistory } from 'react-router-dom';
import { CommandBar, Label } from '@fluentui/react';
import { ResourceKeys } from '../../../../../localization/resourceKeys';
import { getComponentNameFromQueryString } from '../../../../shared/utils/queryStringHelper';
import { DevicePropertiesPerInterface } from './devicePropertiesPerInterface';
import { REFRESH, NAVIGATE_BACK } from '../../../../constants/iconNames';
import { MultiLineShimmer } from '../../../../shared/components/multiLineShimmer';
import { usePnpStateContext } from '../../context/pnpStateContext';
import { SynchronizationStatus } from '../../../../api/models/synchronizationStatus';
import { generateReportedTwinSchemaAndInterfaceTuple } from './dataHelper';
import { dispatchGetTwinAction, getBackUrl } from '../../utils';
import { AppInsightsClient } from '../../../../shared/appTelemetry/appInsightsClient';
import { TELEMETRY_PAGE_NAMES } from '../../../../../app/constants/telemetry';

export const DeviceProperties: React.FC = () => {
    const { t } = useTranslation();
    const { pathname, search } = useLocation();
    const history = useHistory();
    const componentName = getComponentNameFromQueryString(search);

    const { pnpState, dispatch, getModelDefinition } = usePnpStateContext();
    const isLoading = pnpState.twin.synchronizationStatus === SynchronizationStatus.working ||
    pnpState.modelDefinitionWithSource.synchronizationStatus === SynchronizationStatus.working;
    const modelDefinitionWithSource = pnpState.modelDefinitionWithSource.payload;
    const modelDefinition = modelDefinitionWithSource && modelDefinitionWithSource.modelDefinition;
    const extendedModelDefinition = modelDefinitionWithSource && modelDefinitionWithSource.extendedModel;
    const twin = pnpState.twin.payload;
    const twinAndSchema = React.useMemo(() => {
        return generateReportedTwinSchemaAndInterfaceTuple(extendedModelDefinition || modelDefinition, twin, componentName);
    },                                  [twin, modelDefinition, extendedModelDefinition]);

    const renderProperties = () => {
        return (
            <>
                {!twinAndSchema || twinAndSchema.length === 0 ?
                    <Label className="no-pnp-content">{t(ResourceKeys.deviceProperties.noProperties, {componentName})}</Label> :
                    <DevicePropertiesPerInterface twinAndSchema={twinAndSchema} />
                }
            </>
        );
    };

    const handleRefresh = () => {
        dispatchGetTwinAction(search, dispatch);
        getModelDefinition();
    };

    const handleClose = () => {
        const path = pathname.replace(/\/ioTPlugAndPlayDetail\/properties\/.*/, ``);
        history.push(getBackUrl(path, search));
    };

    React.useEffect(() => {
        AppInsightsClient.getInstance()?.trackPageView({name: TELEMETRY_PAGE_NAMES.PNP_PROPERTIES});
    }, []); // tslint:disable-line: align

    if (isLoading) {
        return  <MultiLineShimmer/>;
    }

    return (
        <>
            <CommandBar
                    className="command"
                    items={[
                        {
                            ariaLabel: t(ResourceKeys.deviceProperties.command.refresh),
                            iconProps: {iconName: REFRESH},
                            key: REFRESH,
                            name: t(ResourceKeys.deviceProperties.command.refresh),
                            onClick: handleRefresh
                        }
                    ]}
                    farItems={[
                        {
                            ariaLabel: t(ResourceKeys.deviceProperties.command.close),
                            iconProps: {iconName: NAVIGATE_BACK},
                            key: NAVIGATE_BACK,
                            name: t(ResourceKeys.deviceProperties.command.close),
                            onClick: handleClose
                        }
                    ]}
            />
            {renderProperties()}
        </>
    );

};
