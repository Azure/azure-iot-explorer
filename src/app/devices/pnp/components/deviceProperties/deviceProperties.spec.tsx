/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import 'jest';
import * as React from 'react';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { CommandBar, Shimmer } from '@fluentui/react';
import { DeviceProperties } from './deviceProperties';
import { InterfaceNotFoundMessageBar } from '../../../shared/components/interfaceNotFoundMessageBar';
import { TwinWithSchema } from './dataHelper';
import { PnpStateInterface, pnpStateInitial } from '../../state';
import * as PnpContext from '../../context/pnpStateContext';
import { SynchronizationStatus } from '../../../../api/models/synchronizationStatus';
import { testModelDefinition, testTwin } from './testData';
import { ModelDefinition } from '../../../../api/models/modelDefinition';
import { getDeviceTwinAction } from '../../actions';

const interfaceId = 'urn:contoso:com:EnvironmentalSensor;1';
const pathname = `/#/devices/deviceDetail/ioTPlugAndPlay/ioTPlugAndPlayDetail/properties/?id=device1&componentName=foo&interfaceId=${interfaceId}`;

jest.mock('react-router-dom', () => ({
    useHistory: () => ({ push: jest.fn() }),
    useLocation: () => ({ search: `?id=device1&componentName=foo&interfaceId=${interfaceId}` , pathname }),
}));

export const twinWithSchema: TwinWithSchema = {
    propertyModelDefinition: {
        '@type': 'Property',
        'description': 'The state of the device. Two states online/offline are available',
        'displayName': 'Device State',
        'name': 'state',
        'schema': 'boolean'
    },
    propertySchema: {
        description: 'Device State / The state of the device. Two states online/offline are available',
        required: null,
        title: 'state',
        type: 'boolean'
    },
    reportedTwin: null
};

describe('components/devices/deviceProperties', () => {

    const mockFetchedState = (model: ModelDefinition) => {
        const pnpState: PnpStateInterface = {
            ...pnpStateInitial(),
            twin: {
                payload: testTwin,
                synchronizationStatus: SynchronizationStatus.fetched
            },
            modelDefinitionWithSource: {
                payload: {
                    isModelValid: true,
                    modelDefinition: model,
                    source: null,
                },
                synchronizationStatus: SynchronizationStatus.fetched
            }
        };
        jest.spyOn(PnpContext, 'usePnpStateContext').mockReturnValueOnce({pnpState, dispatch: jest.fn(), getModelDefinition: jest.fn()});
    };

    it('shows Shimmer while loading', () => {
        const pnpState: PnpStateInterface = {
            ...pnpStateInitial(),
            twin: {
                synchronizationStatus: SynchronizationStatus.working
            },
            modelDefinitionWithSource: {
                synchronizationStatus: SynchronizationStatus.fetched
            }
        };
        jest.spyOn(PnpContext, 'usePnpStateContext').mockReturnValueOnce({pnpState, dispatch: jest.fn(), getModelDefinition: jest.fn()});
        const wrapper = mount(<DeviceProperties/>);
        expect(wrapper.find(Shimmer)).toBeDefined();
    });

    it('matches snapshot while interface cannot be found', () => {
        mockFetchedState(null);
        expect(shallow(<DeviceProperties/>)).toMatchSnapshot();
        expect(shallow(<DeviceProperties/>).find(InterfaceNotFoundMessageBar)).toBeDefined();
    });

    it('matches snapshot with one twinWithSchema', () => {
        mockFetchedState(testModelDefinition);
        expect(shallow(<DeviceProperties/>)).toMatchSnapshot();
    });

    it('dispatch action when refresh button is clicked', () => {
        mockFetchedState(testModelDefinition);
        const mockGetDigitalTwinActionSpy = jest.spyOn(getDeviceTwinAction, 'started');
        const wrapper = shallow(<DeviceProperties/>);
        const commandBar = wrapper.find(CommandBar).first();

        act(() => commandBar.props().items[0].onClick(null));
        wrapper.update();
        expect(mockGetDigitalTwinActionSpy).toBeCalled();
    });
});
