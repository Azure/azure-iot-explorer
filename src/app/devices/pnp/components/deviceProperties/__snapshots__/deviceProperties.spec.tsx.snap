// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/devices/deviceProperties matches snapshot while interface cannot be found 1`] = `
<Fragment>
  <StyledCommandBarBase
    className="command"
    farItems={
      Array [
        Object {
          "ariaLabel": "deviceProperties.command.close",
          "iconProps": Object {
            "iconName": "NavigateBack",
          },
          "key": "NavigateBack",
          "name": "deviceProperties.command.close",
          "onClick": [Function],
        },
      ]
    }
    items={
      Array [
        Object {
          "ariaLabel": "deviceProperties.command.refresh",
          "iconProps": Object {
            "iconName": "Refresh",
          },
          "key": "Refresh",
          "name": "deviceProperties.command.refresh",
          "onClick": [Function],
        },
      ]
    }
  />
  <StyledLabelBase
    className="no-pnp-content"
  >
    deviceProperties.noProperties
  </StyledLabelBase>
</Fragment>
`;

exports[`components/devices/deviceProperties matches snapshot with one twinWithSchema 1`] = `
<Fragment>
  <StyledCommandBarBase
    className="command"
    farItems={
      Array [
        Object {
          "ariaLabel": "deviceProperties.command.close",
          "iconProps": Object {
            "iconName": "NavigateBack",
          },
          "key": "NavigateBack",
          "name": "deviceProperties.command.close",
          "onClick": [Function],
        },
      ]
    }
    items={
      Array [
        Object {
          "ariaLabel": "deviceProperties.command.refresh",
          "iconProps": Object {
            "iconName": "Refresh",
          },
          "key": "Refresh",
          "name": "deviceProperties.command.refresh",
          "onClick": [Function],
        },
      ]
    }
  />
  <DevicePropertiesPerInterface
    twinAndSchema={
      Array [
        Object {
          "propertyModelDefinition": Object {
            "@type": "Property",
            "description": "Providing model and optional interfaces information on a digital twin.",
            "displayName": "Model Information",
            "name": "modelInformation",
            "schema": Object {
              "@type": "Object",
              "fields": Array [
                Object {
                  "name": "modelId",
                  "schema": "string",
                },
                Object {
                  "name": "interfaces",
                  "schema": Object {
                    "@type": "Map",
                    "mapKey": Object {
                      "name": "name",
                      "schema": "string",
                    },
                    "mapValue": Object {
                      "name": "schema",
                      "schema": "string",
                    },
                  },
                },
              ],
            },
          },
          "propertySchema": Object {
            "definitions": Object {},
            "properties": Object {
              "interfaces": Object {
                "additionalProperties": Object {
                  "required": Array [],
                  "type": "string",
                },
                "description": "key's name: name, value's name: schema",
                "title": "interfaces",
                "type": "object",
              },
              "modelId": Object {
                "required": Array [],
                "title": "modelId",
                "type": "string",
              },
            },
            "required": Array [],
            "title": "modelInformation",
            "type": "object",
          },
          "reportedTwin": undefined,
        },
        Object {
          "propertyModelDefinition": Object {
            "@type": Array [
              "Property",
              "SemanticType/Brightness Level",
            ],
            "description": "The brightness level for the light on the device. Can be specified as 1 (high), 2 (medium), 3 (low)",
            "displayName": "Brightness Level",
            "name": "brightness",
            "schema": "long",
          },
          "propertySchema": Object {
            "definitions": Object {},
            "required": Array [],
            "title": "brightness",
            "type": Array [
              "number",
              "null",
            ],
          },
          "reportedTwin": undefined,
        },
      ]
    }
  />
</Fragment>
`;
